spirv_hlsl.cpp

```diff
@@ -2164,6 +2309,134 @@ void CompilerHLSL::emit_texture_size_variants(uint64_t variant_mask, const char
 	}
 }
 
+void CompilerHLSL::analyze_meshlet_writes()
+{
+	if (get_execution_model() == ExecutionModelMeshEXT)
+	{
+		uint32_t id_per_vertex = 0;
+		uint32_t id_per_primitive = 0;
+		uint32_t id_payload = 0;
+		ir.for_each_typed_id<SPIRVariable>([&](uint32_t id, SPIRVariable &var) {
+			auto &type = this->get<SPIRType>(var.basetype);
+			bool block = has_decoration(type.self, DecorationBlock);
+			if (var.storage == StorageClassOutput && block && is_builtin_variable(var))
+			{
+				auto flags = get_buffer_block_flags(var.self);
+				if (flags.get(DecorationPerPrimitiveEXT))
+					id_per_primitive = var.self;
+				else
+					id_per_vertex = var.self;
+			}
+			if (var.storage == StorageClassTaskPayloadWorkgroupEXT)
+			{
+				id_payload = var.self;
+			}
+		});
+
+		unordered_set<uint32_t> processed_func_ids;
+		analyze_meshlet_writes(ir.default_entry_point, id_payload, id_per_vertex, id_per_primitive, processed_func_ids);
+	}
+}
+
+void CompilerHLSL::analyze_meshlet_writes(uint32_t func_id, uint32_t id_payload,
+                                          const uint32_t id_per_vertex, const uint32_t id_per_primitive,
+                                          std::unordered_set<uint32_t>& processed_func_ids)
+{
+	// Avoid processing a function more than once
+	if (processed_func_ids.find(func_id) != processed_func_ids.end())
```

Comment:
It seems like this is reimplementing traverse_all_reachable_opcode().

Reply:
Yes, - I've decided to use recursive version: it shorter and more explicit.